python/Ganga/Core/GangaRepository/GangaRepositoryXML.py

```diff
@@ -78,76 +78,104 @@ def get_backupFile(input_filename):
 
     return str(input_filename)+"_"+str(count)
 
+global_lock = threading.Lock()
+
 def safe_save(fn, _obj, to_file, ignore_subs=''):
 
-    # Add a global lock to make absolutely sure we don't have multiple threads writing files
-    # See Issue 185
-    with safe_save.lock:
+    obj = stripProxy(_obj)
 
-        obj = stripProxy(_obj)
+    ## A global save works here but can cause things to slow down.
+    ## This function should be safe on a per Repo per object level,
+    ## otherwise we have multiple objects overwriting each others XML
 
-        check_app_hash(obj)
+    ## Want to make sure that the class is locked as don't want to be performing multiple
+    ## sets of file ops on the same object at the same time
+    if isinstance(obj, GangaObject):
+        parent_lock = GangaObject._getParentLock(obj)
+    else:
+        parent_lock = global_lock
 
-        if not os.path.exists(fn):
-            # file does not exist, so make it fast!
-            try:
-                dirname = os.path.dirname(fn)
-                if not os.path.exists(dirname):
-                    os.makedirs(dirname)
-                if not os.path.isfile(fn):
-                    new_file = open(fn, 'a')
-                    new_file.close()
+    was_locked = False
+    if not parent_lock.locked() and isinstance(obj, GangaObject):
+        parent_lock.acquire()
+        was_locked = True
+    elif not isinstance(obj, GangaObject):
+        parent_lock.acquire()
+        was_locked = True
```

Comment:
Is that in Mark's new test? I found that too and I think it's better fix one small test that change a core function like this to allow for it. I've got a PR incoming to change Mark's `TestSafeSave` test.